### PR TITLE
IA-1433 - add retries to IAM calls in Leo and IA-1434 - clusters stuck in Deleting state

### PR DESCRIPTION
**Pull request by rtitle**
https://broadworkbench.atlassian.net/browse/IA-1433
https://broadworkbench.atlassian.net/browse/IA-1434

\<your comments for this PR go here\>

Have you read [CONTRIBUTING.md](https://github.com/DataBiosphere/leonardo/blob/develop/CONTRIBUTING.md) lately? If not, do that first.

I, the developer opening this PR, do solemnly pinky swear that:

- [x] I've documented my API changes in Swagger

In all cases:

- [ ] Get a thumbsworth of review and PO signoff if necessary
- [ ] Verify all tests go green
- [ ] Squash and merge; you can delete your branch after this
- [ ] Test this change deployed correctly and works on dev environment after deployment

### CODE REVIEW

**Comment by rtitle** [comment]
Travis is acting weird, failing to install oracle jdk..

File: http/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/service/LeonardoServiceSpec.scala [hunk 1]
```diff
@@ -494,17 +494,41 @@ class LeonardoServiceSpec extends TestKit(ActorSystem("leonardotest")) with Flat
...
         // check that the cluster no longer exists
         gdDAO.clusters should not contain key(clusterName)
+
+        // cluster should be in Deleting state (the cluster monitor transitions it to Deleted)
+        dbFutureValue { _.clusterQuery.getClusterStatus(cluster.id) } shouldBe Some(ClusterStatus.Deleting)
     }
   }
 
+  it should "delete a cluster that has status Error and does not have a google id" in isolatedDbTest {
+    //we meed to use a special version of the MockGoogleIamDAO to simulate an error when adding IAM roles
```
Comment (kyuksel) [inline http/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/service/LeonardoServiceSpec.scala]:
```suggestion
    //we need to use a special version of the MockGoogleIamDAO to simulate an error when adding IAM roles
```

File: http/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/service/LeonardoServiceSpec.scala [hunk 2]
```diff
@@ -494,17 +494,41 @@ class LeonardoServiceSpec extends TestKit(ActorSystem("leonardotest")) with Flat
...
 
         // check that the cluster no longer exists
         gdDAO.clusters should not contain key(clusterName)
+
+        // cluster should be in Deleting state (the cluster monitor transitions it to Deleted)
+        dbFutureValue { _.clusterQuery.getClusterStatus(cluster.id) } shouldBe Some(ClusterStatus.Deleting)
     }
   }
 
+  it should "delete a cluster that has status Error and does not have a google id" in isolatedDbTest {
```
Comment (kyuksel) [inline http/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/service/LeonardoServiceSpec.scala:511]: Where are we setting up the cluster to not have a google id?

Comment (rtitle) [reply http/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/service/LeonardoServiceSpec.scala:511]: If an async step fails before the Google creation (such as IAM policies), the cluster will end up in Error state without a Google ID. Due to this bug, clusters in this state were undeletable.

Comment (kyuksel) [reply http/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/service/LeonardoServiceSpec.scala:511]: Right. Do we not wanna make sure the cluster we use for testing has no Google ID to make sure the test is indeed verifying the bug fix?

Comment (rtitle) [reply http/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/service/LeonardoServiceSpec.scala:511]: Yeah good idea.. will add.

Comment (rtitle) [reply http/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/service/LeonardoServiceSpec.scala:511]: Added a verification for this.

**Review by kyuksel — APPROVED** [review]
👍

**Pull request merged**